### PR TITLE
memory leaks and warnings

### PR DESCRIPTION
**Pull request by JohnLangford**
Fixes several memory leaks and warnings.

### CODE REVIEW

File: vowpalwabbit/cbify.cc [hunk 1]
```diff
@@ -88,15 +88,19 @@ void finish(cbify& data)
...
+
   if (data.use_adf)
   {
     for (size_t a = 0; a < data.adf_data.num_actions; ++a)
-    {
-      data.adf_data.ecs[a]->pred.a_s.delete_v();
-      VW::dealloc_example(CB::cb_label.delete_label, *data.adf_data.ecs[a]);
-      free_it(data.adf_data.ecs[a]);
-    }
+      {
```
Comment (lokitoth) [inline vowpalwabbit/cbify.cc]: [Minor][Nit] Is the extra indentation intended?

File: vowpalwabbit/cbify.cc [hunk 2]
```diff
@@ -467,6 +477,7 @@ base_learner* cbifyldf_setup(options_i& options, vw& all)
 
   data->app_seed = uniform_hash("vw", 2, 0);
   data->all = &all;
+  data->use_adf=true;
```
Comment (lokitoth) [inline vowpalwabbit/cbify.cc]: Out of curiosity, what is the impact of not setting this?

File: vowpalwabbit/parse_example_json.h [hunk 3]
```diff
@@ -314,11 +314,8 @@ struct LabelState : BaseState<audit>
 
   BaseState<audit>* StartObject(Context<audit>& ctx) { return ctx.label_object_state.StartObject(ctx); }
 
-  BaseState<audit>* String(Context<audit>& ctx, const char* str, rapidjson::SizeType /* len */, bool copy)
```
Comment (lokitoth) [inline vowpalwabbit/parse_example_json.h]: I think this copy is part of the API required by the rapidjson parser: http://rapidjson.org/md_doc_sax.html#CustomDataStructure. 

I am guessing this is what is causing the parser failures in the unit-tests.

File: vowpalwabbit/parser.cc [hunk 4]
```diff
@@ -999,6 +999,17 @@ void free_parser(vw& all)
     output->currentname.delete_v();
   }
 
+  while (! all.p->example_pool.empty())
```
Comment (lokitoth) [inline vowpalwabbit/parser.cc]: Is it possible to get into a weird state where parser is deallocated, but there is still a thread trying to pass some of these through learn?

(Do we define that as improper consumption of the library?)

File: vowpalwabbit/parse_example_json.h (hunk 3 above)
Comment (lokitoth) [reply vowpalwabbit/parse_example_json.h]: Looks like you were way ahead of me.

File: vowpalwabbit/cbify.cc (hunk 1 above)
Comment (JohnLangford) [reply vowpalwabbit/cbify.cc]: No.  

The problem with styles, in my experience, is that every editor has a different interpretation of precisely what a style is...

File: vowpalwabbit/cbify.cc (hunk 2 above)
Comment (JohnLangford) [reply vowpalwabbit/cbify.cc]: The finish() doesn't work properly.

File: vowpalwabbit/parser.cc (hunk 4 above)
Comment (JohnLangford) [reply vowpalwabbit/parser.cc]: It shouldn't happen given the way shutdown occurs.

**Comment by JohnLangford** [comment]
There's a not-locally-reproducible problem in dsjson tests.  Ideas?

File: vowpalwabbit/parser.cc (hunk 4 above)
Comment (jackgerrits) [reply vowpalwabbit/parser.cc]: Long term it would be better if the pool took a deallocate function and used that when its destructor is called

**Pull request merged**